service/common/src/main/java/org/apache/polaris/service/config/ReservedProperties.java

```diff
@@ -0,0 +1,127 @@
+import org.apache.iceberg.MetadataUpdate;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Used to track entity properties reserved for use by the catalog. These properties may not be
+ * overridden by the end user.
+ */
+public interface ReservedProperties {
+  Logger LOGGER = LoggerFactory.getLogger(ReservedProperties.class);
+
+  /**
+   * A list of prefixes that are considered reserved. Any property starting with one of these
+   * prefixes is a reserved property.
+   */
+  List<String> reservedPrefixes();
+
+  /** If true, attempts to modify a reserved property should throw an exception. */
+  default boolean shouldThrow() {
+    return true;
+  }
+
+  /**
+   * Removes reserved properties from a planned change to an entity. If `shouldThrow`returns true,
+   * this will throw an IllegalArgumentException.
+   *
+   * @param existingProperties The properties currently present for an entity
+   * @param updateProperties The properties present in an update to an entity
+   * @return The keys from the new key list which are not reserved properties
+   */
+  default Map<String, String> removeReservedPropertiesFromUpdate(
+      Map<String, String> existingProperties, Map<String, String> updateProperties)
+      throws IllegalArgumentException {
+    Map<String, String> updatePropertiesWithoutReservedProperties =
+        removeReservedProperties(updateProperties);
+    for (var entry : updateProperties.entrySet()) {
+      // If a key was removed from the update, we substitute in the existing value as to not remove
+      // it
+      if (!updatePropertiesWithoutReservedProperties.containsKey(entry.getKey())) {
+        if (existingProperties.containsKey(entry.getKey())) {
+          updatePropertiesWithoutReservedProperties.put(
+              entry.getKey(), existingProperties.get(entry.getKey()));
+        }
+      }
+    }
+    return updatePropertiesWithoutReservedProperties;
+  }
+
+  /**
+   * Removes reserved properties from a list of input property keys. If `shouldThrow`returns true,
+   * this will throw an IllegalArgumentException.
+   *
+   * @param properties A map of properties to remove reserved properties from
+   * @return The keys from the input list which are not reserved properties
+   */
+  default Map<String, String> removeReservedProperties(Map<String, String> properties)
+      throws IllegalArgumentException {
+    Map<String, String> results = new HashMap<>();
+    List<String> prefixes = reservedPrefixes();
+    for (var entry : properties.entrySet()) {
+      boolean isReserved = false;
+      for (String prefix : prefixes) {
+        if (entry.getKey().startsWith(prefix)) {
+          String message =
+              String.format("Property '%s' matches reserved prefix '%s'", entry.getKey(), prefix);
+          if (shouldThrow()) {
+            throw new IllegalArgumentException(message);
+          } else {
+            LOGGER.debug(message);
+            isReserved = true;
```

Comment:
nit: it would be clearer to move this above line 89 and do a `break` here.

Reply:
Moved the line, but I think we shouldn't break so that multiple logs can be emitted

Reply:
Fair enough.